File: src/main/java/eu/execom/fabut/util/ReflectionUtil.java
package eu.execom.fabut.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import junit.framework.AssertionFailedError;
import eu.execom.fabut.enums.AssertableType;
import eu.execom.fabut.exception.CopyException;

/**
 * Util class for reflection logic needed by testutil.
 * 
 * @author Dusko Vesin
 * @author Nikola Olah
 * @author Bojan Babic
 * @author Nikola Trkulja
 */
public final class ReflectionUtil {

	public static final String SETTER_SUFIX = "_$eq";
	public static final String ID = "id";

	private ReflectionUtil() {
		super();
	}

	/**
	 * Gets field name from setter.
	 * 
	 * @param fromMethod
	 * @return
	 */
	public static String getFieldNameFromSetter(Method fromMethod) {
		return "_"
				+ fromMethod.getName().substring(0,
						fromMethod.getName().indexOf(SETTER_SUFIX));
	}

	/**
	 * Gets field name from getter, ensures that field actually exists. Does
	 * three checks as there are 3 ways of naming fields atm.
	 * 
	 * @param fromMethod
	 * @return
	 */
	public static String getFieldNameFromGetter(Method fromMethod) {
		Class<?> ownerClass = fromMethod.getDeclaringClass();
		try {
			Field field = ownerClass.getDeclaredField(fromMethod.getName());
			return field.getName();
		} catch (Exception e) {
		}

		try {
			Field field = ownerClass.getDeclaredField("_"
					+ fromMethod.getName());
			return field.getName();
		} catch (Exception e) {
		}

		try {
			Field field = ownerClass
					.getDeclaredField(getRubbishScalaPrefix(ownerClass) + "_"
							+ fromMethod.getName());
			return field.getName();
		} catch (Exception e) {
		}

		return null;
	}

	/**
	 * Searches trough property class inheritance tree for field with specified
	 * name. Starting from property class method recursively climbs higher in
	 * the inheritance tree until it finds field with specified name or reached
	 * object in which case returns null.
	 * 
	 * @param fieldClass
	 *            class of the field.
	 * @param fieldName
	 *            name of the field
	 * @return {@link Field} with specified name, otherwise <code>null</code>>
	 */
	public static Field findField(Class<?> fieldClass, String fieldName) {
		if (fieldClass == null) {
			return null;
		}
		try {
			return fieldClass.getDeclaredField(fieldName);
		} catch (final Exception e) {
		}
		// try to find field with scala rubbish prefix
		try {
			return fieldClass.getDeclaredField(ReflectionUtil
					.getRubbishScalaPrefix(fieldClass) + fieldName);
		} catch (final Exception e) {
		}
		return findField(fieldClass.getSuperclass(), fieldName);

	}

	/**
	 * Is specified method for specified class a get method.
	 */
	public static boolean isGetMethod(Class<?> clazz, Method method) {
		String fieldName = getFieldNameFromGetter(method);
		Field field = findField(clazz, fieldName);
		return field != null && method.getReturnType().equals(field.getType());

	}

	/**
	 * Gets all fields that need to be asserted within given object.
	 * 
	 * @param object
	 * @return
	 */
	public static Map<String, Object> getFieldsForAssertFromMethods(
			final Object object) {
		final Map<String, Object> fieldsForAssert = new HashMap<String, Object>();
		final Method[] allMethods = object.getClass().getMethods();
		for (final Method method : allMethods) {
			if (isGetMethod(object.getClass(), method)) {
				try {
					final Object value = method.invoke(object);
					fieldsForAssert.put(getFieldNameFromGetter(method), value);
				} catch (final Exception e) {
				}
			}
		}
		return fieldsForAssert;
	}

	/**
	 * Gets field value from given object for given field name.
	 * 
	 * @param object
	 * @param fieldName
	 * @return
	 * @throws Exception
	 */
	public static Object getFieldValue(Object object, String fieldName)
			throws Exception {
		try {
			int indexOfUnderScore = fieldName.lastIndexOf("_");
			Method getter = object.getClass().getMethod(
					fieldName.substring(indexOfUnderScore + 1));
			return getter.invoke(object);
		} catch (Exception e) {
		}
		return null;
	}

	/**
	 * Gets assertable type from expected/actual pair.
	 * 
	 * @param expected
	 * @param actual
	 * @param types
	 * @return
	 */
	public static AssertableType getAssertableTypeFrom(Object expected,
			Object actual, Map<AssertableType, List<Class<?>>> types) {

		if (expected == null && actual == null) {
			return AssertableType.PRIMITIVE_TYPE;
		}

		final Class<?> typeClass = actual != null ? actual.getClass()
				: expected.getClass();

		if (scala.collection.immutable.List.class.isAssignableFrom(typeClass)) {
			return AssertableType.SCALA_LIST_TYPE;
		} else if (scala.collection.immutable.Map.class
				.isAssignableFrom(typeClass)) {
			return AssertableType.SCALA_MAP_TYPE;

		} else if (types.get(AssertableType.COMPLEX_TYPE).contains(typeClass)) {
			return AssertableType.COMPLEX_TYPE;
		} else if (types.get(AssertableType.ENTITY_TYPE).contains(typeClass)) {
			return AssertableType.ENTITY_TYPE;
		} else if (types.get(AssertableType.IGNORED_TYPE).contains(typeClass)) {
			return AssertableType.IGNORED_TYPE;
		} else {
			return AssertableType.PRIMITIVE_TYPE;
		}
	}

	/**
	 * Creates copy of an given object
	 * 
	 * @param object
	 * @return
	 * @throws CopyException
	 */
	public static Object createCopy(Object object) throws CopyException {
		for (Constructor<?> constructor : object.getClass().getConstructors()) {
			if (constructor.getParameterTypes().length == 1
					&& constructor.getParameterTypes()[0] == object.getClass()) {
				try {
					return constructor.newInstance(object);
				} catch (Exception e) {
					throw new CopyException(object.getClass().getName());
				}
			}
		}
		throw new CopyException(object.getClass().getName());
	}

	/**
	 * Gets id value from given entity.
	 * 
	 * @param entity
	 * @return
	 */
	public static Object getIdValue(Object entity) {
		try {
			Method method = entity.getClass().getMethod(ID);
			return method.invoke(entity);
		} catch (Exception e) {
			return null;
		}
	}

	/**
	 * In some cases scala compiler appends "rubbish" prefix to field name. This
	 * method return "rubbish" prefix from for given class.
	 * 
	 * @param clazz
	 * @return
	 */
	public static String getRubbishScalaPrefix(Class<?> clazz) {
		return clazz.getName().replaceAll("\\.", "\\$") + "$$";
	}

	/**
	 * Checks if given object is instance of scala.Some or scala.None.
	 * 
	 * @param actual
	 */
	public static void checkIfOption(Object actual) {
		if (actual != null
				&& (actual.getClass().equals(scala.Some.class) || actual
						.getClass().equals(scala.None.class))) {
			throw new AssertionFailedError(
					"Object is option, please pass option value to assert!");
		}
	}

	/**
	 * Fabut cannot assert scala lists or maps.
	 * 
	 * @param actual
	 */
	public static void checkIfListOrMap(Object actual) {
		if (actual != null
				&& scala.collection.immutable.List.class
						.isAssignableFrom(actual.getClass())) {
			throw new AssertionFailedError("Lists cannot be asserted!");
		}
		if (actual != null
				&& scala.collection.immutable.Map.class.isAssignableFrom(actual
						.getClass())) {
			throw new AssertionFailedError("Lists cannot be asserted!");
		}
	}
}

File: src/main/java/eu/execom/fabut/util/ReflectionUtil.scala
package eu.execom.fabut.util

import eu.execom.fabut.FabutObjectAssert
import eu.execom.fabut.enums.AssertableType._
import eu.execom.fabut.exception.CopyException
import eu.execom.fabut.graph.NodesList
import eu.execom.fabut.property.Property

import scala.collection.mutable.{ListBuffer, Map => MutableMap}
import scala.reflect.runtime.universe._

/**
 * Util class for reflection logic needed by testutil.
 */
object ReflectionUtil {

  lazy val classLoaderMirror = runtimeMirror(getClass.getClassLoader)
  lazy val SETTER_POSTFIX = "_$eq"

  var fabutAssert: FabutObjectAssert = null

  /**
   * Returns a type of given value
   *
   * @param value
   * @return
   * one of assertable types
   */
  def getAssertableType(value: Any) = value match {
    case _: List[_] => SCALA_LIST_TYPE
    case _: Map[_, _] => SCALA_MAP_TYPE
    case _ if getObjectType(value, COMPLEX_TYPE).isDefined => COMPLEX_TYPE
    case _ if getObjectType(value, ENTITY_TYPE).isDefined => ENTITY_TYPE
    case _ if getObjectType(value, IGNORED_TYPE).isDefined => IGNORED_TYPE
    case _ => PRIMITIVE_TYPE
  }


  /**
   * Returns object type from 'types' for given value
   *
   * @param objectValue
   * object for which we check the type
   * @param assertableType
   * the list of predefined objects from map of types where we should search for value
   * @return
   * the specific type of object
   */

  def getObjectType(objectValue: Any, assertableType: AssertableType): Option[Type] =
    if (objectValue == null) {
      None
    } else {
      try {
        fabutAssert.types(assertableType).find(typeName =>
          typeName.toString == objectValue.getClass.getCanonicalName)

      } catch {
        case e: NoSuchElementException => None
      }
    }


  /**
   * Gets all the properties of given object that need to be asserted.
   *
   * @param objectInstance
   * object instance of fields
   * @param pathName
   * @param objectTypeOption
   * optioned object instance type
   *
   * @return
   * optioned map that has all the properties of object instance
   *
   * @throws ScalaReflectionException
   *
   */
  def getObjectProperties(objectInstance: Any, classTypeOption: Option[Type]): Map[String, Property] = {

    val result: scala.collection.mutable.Map[String, Property] = MutableMap()

    if (classTypeOption == None) {
      return result.toMap
    }

    val classType = classTypeOption.get
    val instanceMirror = classLoaderMirror.reflect(objectInstance)
    val allMembers = extractAllGetMethods(classType)

    try {
      result ++= allMembers.map({
        case member: TermSymbol =>
          val name = member.name.toString
          val value = reflectField(member, classType, instanceMirror)
          (name, Property(name, value.get))
      })
    } catch {
      case e: ScalaReflectionException => println("which field failed with reflection")
    }
    result.toMap
  }

  def extractAllGetMethods(classType: Type): List[TermSymbol] = {

    val allMembers = new ListBuffer[TermSymbol]

    val classes = classType.baseClasses.map(clazz => clazz.typeSignature)

    classes.foreach({ clazz =>
      val terms = clazz.members.collect({ case member if member.isTerm => member.asTerm})
      val membersOfClazz = terms.filter({ member: TermSymbol => member.isGetter && isVariable(member, terms.toList)})

      membersOfClazz.foreach({ member => allMembers += member})
    })
    allMembers.toList
  }

  def isVariable(termMember: TermSymbol, termMembers: List[TermSymbol]): Boolean =
    termMembers.exists(member =>
      member.name.toString.contains(termMember.name.toString) && member.isVar)


  /**
   * Used for @method getFieldsForAssertFromMethods
   *
   * returns value from a field name and instance mirror of class
   *
   * @param propertyName
   * property name that needs to be asserted
   * @param im
   * instance mirror of object instance
   * @param objectType
   * type of object instance
   *
   * @return
   * property value
   * @throws ScalaReflectionException
   */
  def reflectField(propertyName: TermSymbol, objectType: Type, instanceMirror: InstanceMirror): Option[Any] = util.Try {
    val field = instanceMirror.reflectMethod(propertyName.asMethod)
    field()
  }.toOption


  /**
   * Gets field value from given object for given field name via getter.
   *
   * @param propertyName
   * @param objectInstance
   * @param expectedObjectType
   *
   * @return
   * property value or None if it throws exception
   *
   * @throws ScalaReflectionException
   */
  def getFieldValueFromGetter(fieldName: String, objectInstance: Any, expectedObjectTypeOption: Option[Type]): Option[Any] =
    if (expectedObjectTypeOption == None)
      None
    else {
      val expectedObjectType = expectedObjectTypeOption.get
      try {
        val instanceMirror = classLoaderMirror.reflect(objectInstance)
        val terms = expectedObjectType.members.collect({ case member if member.isTerm => member.asTerm})
        val idSymbol = terms.find(member => member.name.toString.contains(fieldName) && member.isGetter).getOrElse(throw new NoSuchElementException("Cannot find getter for id"))
        val field = instanceMirror.reflectMethod(idSymbol.asMethod)
        Some(field())
      } catch {
        case t: ScalaReflectionException =>
          None
      }
    }


  /**
   *
   * Sets the field value from given object for given field name via setter.
   *
   * @param fieldName
   * @param newFieldValue
   * @param objectInstance
   * @param objectType
   *
   * @return
   * field option value or none if it throws exception
   */
  def setField(fieldName: String, newFieldValue: Any, objectInstance: Any, objectType: Type): Boolean = {

    val methodName = fieldName + SETTER_POSTFIX

    try {
      val im = classLoaderMirror.reflect(objectInstance)
      val mSymbol = objectType.member(TermName(methodName)).asMethod
      val methodMirror = im.reflectMethod(mSymbol)
      methodMirror(newFieldValue)
      return true
    } catch {
      case e: IllegalArgumentException => println(e.getMessage)
      case e: ScalaReflectionException => println(e.getMessage) // throw new ili sta?
    }
    false
  }

  /**
   * Creates copy of given object with empty properties
   *
   * @param objectInstance
   * @param objectTypeOption
   *
   */
  def createEmptyCopy(objectInstance: Any, objectTypeOption: Option[Type]): Option[Any] = {

    if (objectTypeOption == None) throw new CopyException("") //TODO

    val objectType = objectTypeOption.get
    val classSymbol = objectType.typeSymbol.asClass
    val classMirror = classLoaderMirror.reflectClass(classSymbol)

    val constructorsList = objectType.decl(termNames.CONSTRUCTOR).asTerm.alternatives.collect {
      case constructor: MethodSymbol if constructor.paramLists.head.size == 1 || constructor.paramLists.head.size == 0 =>
        constructor
    }.sortBy { constructor => constructor.paramLists.head.size}

    if (constructorsList.nonEmpty) {
      val constructor = constructorsList.head
      val reflectConstructor = classMirror.reflectConstructor(constructor)

      Some(reflectConstructor())
    } else {
      throw new CopyException("Default or copy constructor is missing")
    }

  }

  def getIdValue(entity: Any): Any = {
    getFieldValueFromGetter("id", entity, getObjectType(entity, ENTITY_TYPE)).orNull
  }

  /**
   * Determines if a given object is instance of case class
   */
  def isCaseClass(instanceObject: Any): Boolean = {
    val typeMirror = runtimeMirror(instanceObject.getClass.getClassLoader)
    val instanceMirror = typeMirror.reflect(instanceObject)
    val symbol = instanceMirror.symbol

    symbol.isCaseClass
  }

  def createCopy(objectInstance: Any): Any =
    if (objectInstance == null) {
      return null
    } else {
      getAssertableType(objectInstance) match {
        case SCALA_LIST_TYPE => copyList(objectInstance.asInstanceOf[List[_]])
        case SCALA_MAP_TYPE => copyMap(objectInstance.asInstanceOf[Map[_, _]])
        case _ => createCopyObject(objectInstance, new NodesList).get
      }
    }

  def createCopyObject(objectInstance: Any, nodesList: NodesList): Option[Any] = {

    var flag = 0 // TODO make falt boolean
    val copy = nodesList.expected(objectInstance.asInstanceOf[AnyRef]).getOrElse({
        flag = 1
        createEmptyCopy(objectInstance, getObjectType(objectInstance, getAssertableType(objectInstance)))
          .getOrElse(throw new CopyException(objectInstance.getClass.getSimpleName))
      })

    if (flag == 0) {
      return Some(copy)
    }

    nodesList.addPair(copy.asInstanceOf[AnyRef], objectInstance.asInstanceOf[AnyRef])
    val fieldsForCopy = getObjectProperties(objectInstance, getObjectType(objectInstance, getAssertableType(objectInstance)))

    fieldsForCopy.foreach {
      field =>
        val copiedProperty = copyProperty(field._2.value, nodesList)
        copyValueTo(objectInstance, field._1, copiedProperty, copy)
    }

    Some(copy)
  }

  def copyProperty(propertyForCopy: Any, nodesList: NodesList): Any = {

    if (propertyForCopy == null) {
      return null
    }
    getAssertableType(propertyForCopy) match {
      case COMPLEX_TYPE => createCopyObject(propertyForCopy, nodesList).get
      case SCALA_LIST_TYPE => copyList(propertyForCopy.asInstanceOf[List[_]])
      case SCALA_MAP_TYPE => copyMap(propertyForCopy.asInstanceOf[Map[_, _]])
      case _ => propertyForCopy
    }
  }

  def copyList(list: List[_]): List[Any] = {
    list.map { property => copyProperty(property, new NodesList)}
  }

  def copyMap(mapForCopy: Map[_, _]): Map[Any, Any] = {
    mapForCopy.map { case (key, value) => (key, copyProperty(value, new NodesList))}
  }

  def isComplexType(objectInstance: Any): Boolean = {
    getAssertableType(objectInstance) == COMPLEX_TYPE
  }

  def copyValueTo(objectInstance: Any, propertyName: String, copiedProperty: Any, copiedObject: Any): Boolean =
    getObjectType(copiedObject, getAssertableType(copiedObject)) match {
      case Some(objectType) => setField(propertyName, copiedProperty, copiedObject, objectType)
      case None => false
    }

}
